From 406ed2e8130f1a91655ac13ddec3cfcf286764ea Mon Sep 17 00:00:00 2001
From: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Date: Thu, 8 Jun 2017 17:15:25 +0300
Subject: [PATCH] Add R-Car and Xen display units

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
---
 kmscube.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/kmscube.c b/kmscube.c
index ca2c26627840..2340583ffb3f 100644
--- a/kmscube.c
+++ b/kmscube.c
@@ -120,6 +120,8 @@ static uint32_t find_crtc_for_connector(const drmModeRes *resources,
 static int init_drm(void)
 {
 	static const char *modules[] = {
+		"rcar-du",
+		"xendrm-du",
 		"exynos",
 		"i915",
 		"msm",
-- 
2.7.4

